### PR TITLE
Add New Integration for DeepInfra LLM Models

### PR DESCRIPTION
**Pull request by ovuruska**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## New Package?

Did I fill in the `tool.llamahub` section in the `pyproject.toml` and provide a detailed README.md for my new integration or package?

- [x] Yes
- [ ] No

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

## Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+        Private method to perform a synchronous streaming request to the DeepInfra API.
+
+        Args:
+            endpoint (str): The API endpoint to send the request to.
+            payload (Dict[str, Any]): The request payload.
+
+        Yields:
+            str: The streaming response from the API.
+        """
+        payload["stream"] = True
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: It is bad practice to modify objects passed to a function (unless this is the core purpose of the fn). In this particular case you can either do {**payload, 'stream': True} or just pass `stream: True` in the outer function.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+        Checks whether the incoming data is an actual
+        SSE data message.
+
+        Args:
+            data (bytes): The incoming data.
+
+        Returns:
+            Union[dict, None]: The decoded data or None.
+        """
+        if data and data.startswith(b"data: "):
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: According to the SPEC there might not be a space after `data:` (if I remember correctly).

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+
+        Returns:
+            Dict[str, Any]: The API response.
+        """
+        headers = {
+            "Authorization": f"Bearer {self._api_key}",
+            "Content-Type": "application/json",
+        }
+        response = requests.post(self.get_url(endpoint), json=payload, headers=headers)
+        response.raise_for_status()
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: There should be an option to pass in the above line to do the `raise_for_status` automatically

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+            is_chat_model=self._is_chat_model,
+            model_name=self.model,
+        )
+
+    @property
+    def _is_chat_model(self) -> bool:
+        return True
+
+    # Synchronous Methods
+    def complete(self, prompt: str, **kwargs) -> CompletionResponse:
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py:161]: So in all those functions `**kwargs` will be properly filled with `temperature` and `additional_kwargs`? This is a weird interface, but if it works...

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+    )
+
+    generate_kwargs: Dict[str, Any] = Field(
+        default_factory=dict, description="Additional keyword arguments for generation."
+    )
+
+    def __init__(
+        self,
+        model: str = DEFAULT_MODEL_NAME,
+        temperature: float = DEFAULT_TEMPERATURE,
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: Is temperature the only `first-class` field in here? What about other implementations?

**Review by ichernev — COMMENTED** [review]
Overall looks good, some notes.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py (hunk 1 above)
Comment (ovuruska) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: I just missed that part.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py (hunk 2 above)
Comment (ovuruska) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: Yes, you are right about that.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py (hunk 3 above)
Comment (ovuruska) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: I searched for this one. This is the legit way.
Chaining is yet another possibility, however, then we lose the response object.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py (hunk 4 above)
Comment (ovuruska) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py:161]: I agree with you. However, even it seems weird and a bit abundant, it allows request-wise flexibility on the parameters.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py (hunk 5 above)
Comment (ovuruska) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: I think we should extend this constructor. You're right about that.

Comment (ichernev) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: Now we have temp and max_tokens... I think we should add most (if not all) of the openai supported ones. The extra kwargs is to help us support newer features that are not (yet) ported to the "official" args. Sort-of like a temporary workaround until we (you :)) add them here.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py (hunk 4 above)
Comment (ichernev) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py:161]: I understand it is flexible, I don't understand who builds the arguments. Does the library somehow build arguments from temp, max_tokens + extra args? Is there a function to return "default args" that users can then tweak per request?

I don't understand how the constructor-passed args fit in the picture.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+    retry_request,
+    aretry_request,
+)
+
+"""DeepInfra API base URL."""
+API_BASE = "https://api.deepinfra.com/v1"
+"""DeepInfra Inference API endpoint."""
+INFERENCE_ENDPOINT = "inference"
+"""Chat API endpoint for DeepInfra."""
+CHAT_API_ENDPOINT = "openai/chat/completions"
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py]: I'd use `/v1/openai/chat/completions` for chat and `/v1/openai/completions` for text. The `v1` is not part of the API_BASE (some of our apis don't have v1, and it is more of a per-path thing, not global.

File: llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+        model: str = DEFAULT_MODEL_NAME,
+        additional_kwargs: Optional[Dict[str, Any]] = None,
+        temperature: float = DEFAULT_TEMPERATURE,
+        max_tokens: Optional[int] = None,
+        max_retries: int = 10,
+        api_base: str = API_BASE,
+        timeout: Optional[float] = None,
+        api_key: Optional[str] = None,
+        callback_manager: Optional[CallbackManager] = None,
+        system_prompt: Optional[str] = None,
```
Comment (ichernev) [inline llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py:111]: I believe this is still not handled

Comment (ovuruska) [reply llama-index-integrations/llms/llama-index-llms-deepinfra/llama_index/llms/deepinfra/base.py:111]: These parameters are used in parent and grandparent classes.

**Pull request merged**